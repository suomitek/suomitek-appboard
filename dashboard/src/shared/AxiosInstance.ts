import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  InternalServerError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response) {
        if (err.response.data.message) {
          message = err.response.data.message;
        }
        if (typeof err.response.data === "string") {
          message = err.response.data;
        }
      }

      const dispatchErrorAndLogout = (m: string) => {
        // Global action dispatch to log the user out
        dispatch(actions.auth.authenticationError(m));
        // Expire the session if we are using OIDC tokens and
        // logout either way.
        dispatch(actions.auth.expireSession());
      };
      const response = err.response as AxiosResponse;
      switch (response && response.status) {
        case 401:
          dispatchErrorAndLogout(message);
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          // A 403 directly from the auth proxy requires reauthentication.
          if (Auth.usingOIDCToken() && Auth.is403FromAuthProxy(response)) {
            dispatchErrorAndLogout(message);
          }
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        case 500:
          return Promise.reject(new InternalServerError(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
export const axiosWithAuth = Axios.create();
