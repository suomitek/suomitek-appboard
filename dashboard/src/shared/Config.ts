import axios from "axios";

interface ICluster {
  name: string;
  apiServiceURL: string;
  certificateAuthorityData?: string;
}

export interface IFeatureFlags {
  operators: boolean;
  ui: string;
}

// IConfig is the configuration for Kubeapps
export interface IConfig {
  namespace: string;
  appVersion: string;
  authProxyEnabled: boolean;
  oauthLoginURI: string;
  oauthLogoutURI: string;
  error?: Error;
  featureFlags: IFeatureFlags;
  clusters: ICluster[];
}

export default class Config {
  public static async getConfig() {
    const url = Config.APIEndpoint;
    const { data } = await axios.get<IConfig>(url);

    // Development environment config overrides
    // TODO(miguel) Rename env variable to KUBEAPPS_NAMESPACE once/if we eject create-react-app
    // Currently we are using REACT_APP_* because it's the only way to inject env variables in a sealed setup.
    // Please note that this env variable gets mapped in the run command in the package.json file
    if (process.env.NODE_ENV !== "production") {
      data.appVersion = "DEVEL";
      if (process.env.REACT_APP_KUBEAPPS_NS) {
        data.namespace = process.env.REACT_APP_KUBEAPPS_NS;
      }
    }

    return data;
  }

  private static APIEndpoint: string = "config.json";
}
