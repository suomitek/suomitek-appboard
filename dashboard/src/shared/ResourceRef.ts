import { filter, matches } from "lodash";
import { Kube } from "./Kube";
import { ResourceKind } from "./ResourceKinds";
import { IClusterServiceVersionCRDResource, IK8sList, IResource } from "./types";

export function fromCRD(
  r: IClusterServiceVersionCRDResource,
  cluster: string,
  namespace: string,
  ownerReference: any,
) {
  const resource = {
    apiVersion: Kube.resourceAPIVersion(r.kind as ResourceKind),
    kind: r.kind,
    metadata: {},
  } as IResource;
  // Avoid namespace for cluster-wide resources supported (ClusterRole, ClusterRoleBinding)
  // TODO(andresmgot): This won't work for new resource types, we would need to dinamically
  // resolve those
  const resourceNamespace = r.kind.startsWith("Cluster") ? "" : namespace;
  const ref = new ResourceRef(resource, cluster, resourceNamespace);
  ref.filter = {
    metadata: { ownerReferences: [ownerReference] },
  };
  return ref;
}

// ResourceRef defines a reference to a namespaced Kubernetes API Object and
// provides helpers to retrieve the resource URL
class ResourceRef {
  public cluster: string;
  public apiVersion: string;
  public kind: ResourceKind;
  public name: string;
  public namespace: string;
  public filter: any;

  // Creates a new ResourceRef instance from an existing IResource. Provide
  // defaultNamespace to set if the IResource doesn't specify a namespace.
  constructor(r: IResource, cluster: string, defaultNamespace?: string) {
    this.cluster = cluster;
    this.apiVersion = r.apiVersion;
    this.kind = r.kind;
    this.name = r.metadata.name;
    this.namespace = r.metadata.namespace || defaultNamespace || "";
    return this;
  }

  // Gets a full resource URL for the referenced resource
  public getResourceURL() {
    return Kube.getResourceURL(
      this.cluster,
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.namespace,
      this.name,
    );
  }

  public watchResourceURL() {
    return Kube.watchResourceURL(
      this.cluster,
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.namespace,
      this.name,
    );
  }

  public async getResource() {
    const resource = await Kube.getResource(
      this.cluster,
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.namespace,
      this.name,
    );
    const resourceList = resource as IK8sList<IResource, {}>;
    if (resourceList.items) {
      resourceList.items = filter(resourceList.items, matches(this.filter));
      return resourceList;
    }
    return resource;
  }

  // Opens and returns a WebSocket for the requested resource. Note: it is
  // important that this socket be properly closed when no longer needed. The
  // returned WebSocket can be attached to an event listener to read data from
  // the socket.
  public watchResource() {
    return Kube.watchResource(
      this.cluster,
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.namespace,
      this.name,
    );
  }
}

export default ResourceRef;
