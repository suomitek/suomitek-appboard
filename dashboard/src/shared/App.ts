import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const KUBEOPS_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static async create(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.list(cluster, namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(
    cluster: string,
    namespace: string,
    releaseName: string,
    revision: number,
  ) {
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(
    cluster: string,
    namespace: string,
    releaseName: string,
    purge: boolean,
  ) {
    let endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  public static async listApps(cluster: string, namespace?: string, allStatuses?: boolean) {
    let endpoint = namespace
      ? url.kubeops.releases.list(cluster, namespace)
      : url.kubeops.releases.listAll(cluster);
    if (allStatuses) {
      endpoint += "?statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(endpoint);
    return data.data;
  }

  public static async getRelease(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      url.kubeops.releases.get(cluster, namespace, name),
    );
    return data.data;
  }
}
