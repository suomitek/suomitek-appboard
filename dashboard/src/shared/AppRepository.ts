import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace, name));
    return data;
  }

  public static async resync(name: string, namespace: string) {
    const repo = await AppRepository.get(name, namespace);
    repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
    repo.spec.resyncRequests++;
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(namespace, name), repo);
    return data;
  }

  public static async update(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(namespace, name),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(name, namespace),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(namespace),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async validate(repoURL: string, authHeader: string, customCA: string) {
    const { data } = await axiosWithAuth.post<any>(url.backend.apprepositories.validate(), {
      appRepository: { repoURL, authHeader, customCA },
    });
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getSelfLink(namespace: string, name?: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
