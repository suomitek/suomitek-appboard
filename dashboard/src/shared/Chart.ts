import { JSONSchema4 } from "json-schema";
import { axiosWithAuth } from "./AxiosInstance";
import { IChart, IChartVersion } from "./types";
import * as URL from "./url";

export default class Chart {
  public static async fetchCharts(namespace: string, repo: string) {
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(
      URL.api.charts.list(namespace, repo),
    );
    return data.data;
  }

  public static async fetchChartVersions(namespace: string, id: string): Promise<IChartVersion[]> {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion[] }>(
      URL.api.charts.listVersions(namespace, id),
    );
    return data.data;
  }

  public static async getChartVersion(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion }>(
      URL.api.charts.getVersion(namespace, id, version),
    );
    return data.data;
  }

  public static async getReadme(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getReadme(namespace, id, version),
    );
    return data;
  }

  public static async getValues(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getValues(namespace, id, version),
    );
    return data;
  }

  public static async getSchema(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<JSONSchema4>(
      URL.api.charts.getSchema(namespace, id, version),
    );
    return data;
  }

  public static async listWithFilters(
    namespace: string,
    name: string,
    version: string,
    appVersion: string,
  ) {
    const url = `${
      Chart.APIEndpoint
    }/ns/${namespace}/charts?name=${name}&version=${encodeURIComponent(
      version,
    )}&appversion=${appVersion}`;
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/assetsvc/v1";
}
