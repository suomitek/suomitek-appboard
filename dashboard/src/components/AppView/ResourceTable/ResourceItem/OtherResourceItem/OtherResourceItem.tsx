import * as React from "react";
import ResourceRef from "../../../../../shared/ResourceRef";
import { IResource } from "../../../../../shared/types";

export interface IOtherResourceProps {
  resource: IResource | ResourceRef;
}

export const OtherResourceColumns: React.SFC = () => {
  return (
    <React.Fragment>
      <th className="col-6">NAME</th>
      <th className="col-6">KIND</th>
    </React.Fragment>
  );
};

const OtherResourceItem: React.SFC<IOtherResourceProps> = props => {
  const { resource } = props;
  let name;
  const fullResource = resource as IResource;
  if (fullResource.metadata && fullResource.metadata.name) {
    name = fullResource.metadata.name;
  } else {
    const resourceRef = resource as ResourceRef;
    name = resourceRef.name;
  }
  return (
    <>
      <td className="col-6">{name}</td>
      <td className="col-6">{resource.kind}</td>
    </>
  );
};

export default OtherResourceItem;
