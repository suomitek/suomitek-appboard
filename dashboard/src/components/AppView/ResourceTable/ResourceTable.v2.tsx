import React, { useEffect, useMemo } from "react";
import { IKubeItem, IResource, ISecret, IStoreState } from "shared/types";

import actions from "actions";
import { CdsIcon } from "components/Clarity/clarity";
import Table from "components/js/Table";
import LoadingWrapper from "components/LoadingWrapper/LoadingWrapper.v2";
import { useDispatch, useSelector } from "react-redux";
import ResourceRef from "shared/ResourceRef";
import { flattenResources } from "shared/utils";
import { DaemonSetColumns } from "./ResourceData/DaemonSet";
import { DeploymentColumns } from "./ResourceData/Deployment";
import { OtherResourceColumns } from "./ResourceData/OtherResource";
import { SecretColumns } from "./ResourceData/Secret";
import { ServiceColumns } from "./ResourceData/Service";
import { StatefulSetColumns } from "./ResourceData/StatefulSet";

interface IResourceTableProps {
  id: string;
  title?: string;
  resourceRefs: ResourceRef[];
  avoidEmptyResource?: boolean;
}

function getColumns(r: ResourceRef) {
  switch (r.kind) {
    case "Deployment":
      return DeploymentColumns;
    case "StatefulSet":
      return StatefulSetColumns;
    case "DaemonSet":
      return DaemonSetColumns;
    case "Service":
      return ServiceColumns;
    case "Secret":
      return SecretColumns;
    default:
      return OtherResourceColumns;
  }
}

function getData(
  name: string,
  accessors: string[],
  getters: Array<(r: any) => string | JSX.Element | JSX.Element[]>,
  resource?: IKubeItem<IResource | ISecret>,
) {
  const data = {
    name,
  };
  if (!resource || resource.isFetching) {
    data[accessors[1]] = <LoadingWrapper small={true} />;
    return data;
  }
  if (resource.error) {
    data[accessors[1]] = (
      <>
        <CdsIcon shape="alert-triangle" />
        Error: {resource.error.message}
      </>
    );
    return data;
  }
  if (resource.item) {
    accessors.forEach((accessor, index) => {
      data[accessor] = getters[index](resource.item);
    });
    return data;
  }
  data[accessors[1]] = <span>Unkown</span>;
  return;
}

function ResourceTable({ id, title, resourceRefs }: IResourceTableProps) {
  const dispatch = useDispatch();
  useEffect(() => {
    resourceRefs.forEach(r => dispatch(actions.kube.getAndWatchResource(r)));
    return function cleanup() {
      resourceRefs.forEach(r => dispatch(actions.kube.closeWatchResource(r)));
    };
  }, [resourceRefs, dispatch]);
  const resources = useSelector((state: IStoreState) =>
    flattenResources(resourceRefs, state.kube.items),
  );

  const columns = useMemo(() => getColumns(resourceRefs[0]), [resourceRefs]);
  const data = useMemo(
    () =>
      resources.map((resource, index) =>
        getData(
          resourceRefs[index].name,
          columns.map(c => c.accessor),
          columns.map(c => c.getter),
          resource,
        ),
      ),
    [columns, resourceRefs, resources],
  );

  let section: JSX.Element | null = null;
  if (resourceRefs.length > 0) {
    section = (
      <section aria-labelledby={`${id}-table`}>
        {title && <h6 id={`${id}-table`}>{title}</h6>}
        <Table columns={columns} data={data} />
      </section>
    );
  }
  return section;
}

export default ResourceTable;
