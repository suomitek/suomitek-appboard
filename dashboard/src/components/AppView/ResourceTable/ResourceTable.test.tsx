import { shallow } from "enzyme";
import * as React from "react";

import ResourceItemContainer from "../../../containers/ResourceItemContainer";
import ResourceRef from "../../../shared/ResourceRef";
import { IResource } from "../../../shared/types";
import OtherResourceItem from "./ResourceItem/OtherResourceItem";
import ResourceTable from "./ResourceTable";

const clusterName = "cluster-name";

it("skips the element if there are no resources", () => {
  const wrapper = shallow(<ResourceTable resourceRefs={[]} title={""} />);
  expect(wrapper.find(ResourceItemContainer)).not.toExist();
  expect(wrapper.html()).toBe(null);
});

it("renders a ResourceItem", () => {
  const resourceRefs = [
    new ResourceRef(
      { kind: "Deployment", metadata: { name: "foo" } } as IResource,
      clusterName,
      "default",
    ),
  ];
  const wrapper = shallow(<ResourceTable resourceRefs={resourceRefs} title={""} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(ResourceItemContainer)).toExist();
});

it("renders two resources", () => {
  const deployRefs = [
    new ResourceRef(
      { kind: "Deployment", metadata: { name: "foo" } } as IResource,
      clusterName,
      "default",
    ),
    new ResourceRef(
      { kind: "Deployment", metadata: { name: "bar" } } as IResource,
      clusterName,
      "default",
    ),
  ];
  const wrapper = shallow(<ResourceTable resourceRefs={deployRefs} title={""} />);
  expect(wrapper.find(ResourceItemContainer).length).toBe(2);
  expect(
    wrapper
      .find(ResourceItemContainer)
      .at(0)
      .prop("resourceRef"),
  ).toBe(deployRefs[0]);
  expect(
    wrapper
      .find(ResourceItemContainer)
      .at(1)
      .prop("resourceRef"),
  ).toBe(deployRefs[1]);
});

it("renders OtherResourceItem", () => {
  const resourceRefs = [
    new ResourceRef(
      { kind: "ConfigMap", metadata: { name: "foo" } } as IResource,
      clusterName,
      "default",
    ),
  ];
  const wrapper = shallow(<ResourceTable resourceRefs={resourceRefs} title={""} />);
  expect(wrapper.find(OtherResourceItem)).toExist();
  expect(wrapper.find(ResourceItemContainer)).not.toExist();
});

it("renders OtherResource as ItemContainer", () => {
  const resourceRefs = [
    new ResourceRef(
      { kind: "ConfigMap", metadata: { name: "foo" } } as IResource,
      clusterName,
      "default",
    ),
  ];
  const wrapper = shallow(
    <ResourceTable resourceRefs={resourceRefs} title={""} requestOtherResources={true} />,
  );
  expect(wrapper.find(OtherResourceItem)).not.toExist();
  expect(wrapper.find(ResourceItemContainer)).toExist();
});
