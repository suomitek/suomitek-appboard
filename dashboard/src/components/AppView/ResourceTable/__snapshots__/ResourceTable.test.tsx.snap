// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a ResourceItem 1`] = `
<Fragment>
  <h6 />
  <table>
    <thead>
      <tr
        className="flex"
      >
        <DeploymentColumns />
      </tr>
    </thead>
    <tbody>
      <Connect(WorkloadItem)
        key="api/clusters/cluster-name/api/v1/namespaces/default/deployments/foo"
        resourceRef={
          ResourceRef {
            "apiVersion": undefined,
            "cluster": "cluster-name",
            "filter": undefined,
            "kind": "Deployment",
            "name": "foo",
            "namespace": "default",
          }
        }
      />
    </tbody>
  </table>
</Fragment>
`;
